[Dashboard][Lens] Add "convert to lens" action to dashboard

VladLasitsa opened the pull request:
## Summary

Closes https://github.com/elastic/kibana/issues/147032
Completes part of: https://github.com/elastic/kibana/issues/144605

Added `convert to lens` action for panel in dashboards.

If legacy visualization can be converted, the notification 'dot' will shown on context menu.
<img width="828" alt="Снимок экрана 2022-12-02 в 10 50 58" src="https://user-images.githubusercontent.com/16915480/205253599-3f3f102e-8fdc-497c-9e81-a9e1a146687c.png">

New action looks like this:
<img width="781" alt="Снимок экрана 2022-12-02 в 10 52 42" src="https://user-images.githubusercontent.com/16915480/205253909-79d65fd8-81d8-4cce-a61a-234d3996cf84.png">

After clicking by that action user will be navigate to lens page and see the following, where user can replace legacy visualization to lens on dashboard:
<img width="1347" alt="Снимок экрана 2022-12-02 в 10 53 23" src="https://user-images.githubusercontent.com/16915480/205254013-6e26b54d-6b92-4da5-be64-01b2876ea847.png">

On save user also can replace panel on dashboard:
<img width="506" alt="Снимок экрана 2022-12-02 в 10 55 22" src="https://user-images.githubusercontent.com/16915480/205254409-163ebf51-c075-4c9a-a070-cebc7001636d.png">

elasticmachine commented:
Pinging @elastic/kibana-visualizations @elastic/kibana-visualizations-external (Team:Visualizations)

elasticmachine commented:
Pinging @elastic/kibana-presentation (Team:Presentation)

VladLasitsa commented:
@elasticmachine merge upstream

jbudz reviewed (approved):
ftr_configs.yml

MichaelMarcialis reviewed (commented):
Thanks for putting this together, @VladLasitsa. This is looking great! I've added my list of initial comments below for your review. Let me know if you have any questions.

#### Dashboard app comments

- There is currently a bug where the accent colored dot appears in top-right corner of compatible visualization panels in view mode. When attempting to interact with the options menu button, the dot then disappears. The dot shouldn't appear at all in view mode.

![image](https://user-images.githubusercontent.com/3884767/205379512-bcdffab6-618e-41cf-82b3-e1f0ccd2f69e.png)

- Can we append the originating app name to the "Edit visualization" action in the options context menu to better match the wireframes (ex. "Edit visualization in TSVB"). I think it helps provide better incite as to where the user will be taken, especially after adding this new Lens conversion option.

- Would we be open to changing the "Clone panel" action in the options context menu to "Duplicate panel" (and also remember to update text in subsequent toast)? I think that might be more consistent with the language we use elsewhere in Kibana.

- Would we be open to changing the "Edit panel title" action in the options context menu to "Customize panel title" (and also remember to update title in subsequent modal)? Also, could we consider changing its `pencil` icon to `lettering`? I think these changes would better differentiate this action from the "Edit visualization" action.

- Would we be open to changing the "More" action in the options context menu to "Additional options" (and also remember to update title in subsequent context menu layer)?

- The "Convert to Lens" action is currently buried in the "More" section (the second layer of the options context menu). As shown in the wireframes, I was hoping it would be placed in the first layer of the options context menu (after "Edit visualization") for easy discoverability and access. Would it be possible to change

![image](https://user-images.githubusercontent.com/3884767/205376639-eeb8a1d6-7359-4ce6-affd-fa7ce229e51f.png)

- Can we change the "Convert to Lens" action in the options context menu to "Convert visualization to Lens" to better match the wireframes?

- Converting a visualization and replacing it doesn't carry forward the original panel's title settings in the conversion (both the title text and whether it was set to be hidden or not).  Would it be possible for us to carry over these panel title settings in the conversion and replacement process to avoid bothering the user? Also, if it is possible to carry these panel title settings over, can we also append ` (converted)` to the panel title text automatically (similar to how we do when saving a converted visualization to the library)?

#### Lens app comments

- Converting a by-value visualization without a panel title shows `Converting "{visualizationType}"` in the breadcrumbs (ex. `Converting "TSVB visualization"`). In these situations, can we remove the quotes (as it's not referencing a specific panel or saved object title)? For example: `Converting TSVB visualization`.

- When converting a by-reference visualization to Lens, we seem to be using only the panel title text for the conversion breadcrumbs and saved object title input (in the save modal). Would it be better to instead use the originating saved object title instead for by-reference conversions?

- In my wireframes, I had suggested adding a divider between the cancel/save/replace actions group in the the top-right toolbar (to visually separate it from actions that exist apart from the conversion flow). It looks like that didn't make it into this PR. Would it be possible to add? We do something similar in TSVB to separate the "Edit visualization in Lens" action.

![image](https://user-images.githubusercontent.com/3884767/205376262-afa3d1db-e667-4ef7-94b7-726ea4c90e24.png)

- Instead of the `checkInCircleFilled` icon for the "Replace in dashboard" button, can we use the `merge` icon to be consistent with the dashboard options menu?

- Currently, after saving a converted visualization to the library in Lens and choosing **NOT** to replace the panel on dashboard, users are given the option to "Save" or "Replace in dashboard". The "Replace in dashboard" action should be changed to "Save and replace" (with accompanying `save` icon) to better match the wireframes and indicate what is actually happening (as the object is saved in the process).

- Can we add a `save` icon to the `Save and return` and `Save` modal buttons to match wireframes?

stratoula reviewed (commented):
The dot on view mode  appears for the by ref convertable visualizations but not for the by value. I think we don't want the dot to appear on the view mode, only on the edit cc @MichaelMarcialis 

<img width="880" alt="image" src="https://user-images.githubusercontent.com/17003240/205587738-fe8ddf4f-c000-4be6-ad92-376120566791.png">

Also as we want to promote it the panel item should be on the first options and not under More.

stratoula commented:
@elastic/kibana-presentation how do you feel about @MichaelMarcialis 's comments https://github.com/elastic/kibana/pull/146363#pullrequestreview-1202995395 for the dashboard context panel?

stratoula commented on x-pack/plugins/lens/public/app_plugin/app.tsx:
```diff
@@ -250,6 +250,17 @@ export function App({
+    if (
+      !persistedDoc?.title &&
+      initialContext &&
+      initialContext.originatingApp === 'dashboards' &&
```
I was hoping that we could omit using "dashboards" as check here. This is also visible in canvas. We need another approach here and I know that @ThomThomson is against it.

VladLasitsa replied:
Removed it

stratoula reviewed (commented):
I see the following bug when I try to do it on canvas
I think due to this we have the following bug. Go to Canvas, add Flights Delay bucket chart. click the gear, click Convert to Lens and Save and Return. Chart is broken

<img width="1031" alt="image" src="https://user-images.githubusercontent.com/17003240/205590965-56d54402-9ed5-43b6-b60b-7f81e18feb56.png">

darnautov commented:
hi @VladLasitsa, changes in the ML plugin seem irrelevant to this PR. Shall we revert? If not, could you please elaborate on why you had to add them?

VladLasitsa commented:
> hi @VladLasitsa, changes in the ML plugin seem irrelevant to this PR. Shall we revert? If not, could you please elaborate on why you had to add them?

Hi @darnautov, After I've added new action which use `reactToUiComponent` from `@kbn/kibana-react-plugin/public`, I saw that some tests failed in ml plugin because of the `reactToUiComponent` was not mocked. Because if that I've added that mock.

darnautov reviewed (approved):
ML changes LGTM

VladLasitsa commented on src/plugins/embeddable/public/lib/panel/_embeddable_panel.scss:
```diff
@@ -106,6 +106,17 @@
+  position: absolute;
+  top: 0;
+  right: 0;
+  content: '•';
```
@MichaelMarcialis, Adding EuiIcon anyway will require styles for correct position. For me in that case using pseudo element is good solution. Could you please describe a reason why is better adding new element instead of using pseudo element?

MichaelMarcialis replied:
The only reason I asked was for the sake of consistency with other areas of Kibana that use the `dot` icon. That consistency would really only be in regard to size in this case, since the icon is a simple circle. That said, it's not a huge deal either way. If positioning an `EuiIcon` instead of a pseudo element is more complex, I'm OK keeping as you have it.

VladLasitsa replied:
@MichaelMarcialis, Could you please tell me where we have the same dot-notification idea in kibana?

MichaelMarcialis replied:
I believe this is the first use of a `dot` icon to draw attention to a menu button in Kibana, but we do utilize the `dot` icon in a variety of other across Kibana as a means to visually offset or grab attention from users. Some that immediately jump to mind are 1) for indicating change/dirty state when making edits on some pages (like user profiles) and 2) for indicating incompatibility of quick functions in Lens.

VladLasitsa commented:
@MichaelMarcialis about that:
> Can we change the "Convert to Lens" action in the options context menu to "Convert visualization to Lens" to better match the wireframes?

Adding 'visualization' word will wrap text on two lines. I don't think that 'visualization' is too required here, it will be understandable for user that this action will convert 'visualization'

VladLasitsa commented:
> I see the following bug when I try to do it on canvas I think due to this we have the following bug. Go to Canvas, add Flights Delay bucket chart. click the gear, click Convert to Lens and Save and Return. Chart is broken
> 
> <img alt="image" width="1031" src="https://user-images.githubusercontent.com/17003240/205590965-56d54402-9ed5-43b6-b60b-7f81e18feb56.png">

The problem appears because canvas don't work as 'edit by value' on dashboard. It means that for correct work when user click on `Save and Return` we should firstly create a new Lens visualization in library (for now that don't work like this). As we can reproduce it on main using `Edit in Lens` button in visualization page (need create an issue and discuss with canvas team), was decided to not show `Convert to Lens` action on Canvas for now. 

cc @stratoula

ThomThomson commented:
@VladLasitsa, Canvas _does_ actually support by value panels! No need to create a library visualization. I tested this out with the existing TSVB -> Lens behaviour with by value panels and it works fine. 

Just a guess but I think what's likely missing is the `incomingEmbeddable` package is not properly sent back to Canvas in this case.

> I find a way to carry title, but didn't find a good solution to carry hidden title setting
It looks like the title and `hide title settings` are also not maintained with the current TSVB -> Lens flow. These settings are part of the `explicitInput` of the panel. 

We do not spread the existing `explicitInput` [when changing types](https://github.com/elastic/kibana/blob/main/src/plugins/dashboard/public/application/lib/build_dashboard_container.ts#L112). This means that you'll have to ensure that both `title` and `hidePanelTitles` are taken from the TSVB input and translated properly when navigating to Lens.

VladLasitsa commented:
> @VladLasitsa, Canvas _does_ actually support by value panels! No need to create a library visualization. I tested this out with the existing TSVB -> Lens behaviour with by value panels and it works fine: If you **start with a by value** TSVB the issue doesn't happen.
> 
> My best guess as to what's happening here is that there is some bug on the Canvas side when the original chart was `by reference` and we are replacing it with a `by value` chart. Probably some logic that assumes that by reference charts don't need their input replaced. We'll look into it on the Canvas side. @cqliu1 FYI. Tracking this bug [here](https://github.com/elastic/kibana/issues/147032)

Thank you @ThomThomson, I find a possible way to fix it, when we change type of visulization we should not provide savedObjectId. We can add check here `x-pack/plugins/canvas/public/components/hooks/workpad/use_incoming_embeddable.ts`. If original embeddable type is not the same as new type use something like this: `updatedInput.savedObjectId = undefined;`. I've checked locally and it was fixed problem. What do you think about that?

stratoula commented:
> @stratoula yes, let's do Dashboard context menu changes in a separate PR. I have added this as a discussion item for @elastic/kibana-presentation next team meeting and we can follow up with making these changes.

Thanx @nreese!

> My best guess as to what's happening here is that there is some bug on the Canvas side when the original chart was by reference and we are replacing it with a by value chart. Probably some logic that assumes that by reference charts don't need their input replaced. We'll look into it on the Canvas side. @cqliu1 FYI. Tracking this bug https://github.com/elastic/kibana/issues/147032

Thanx @ThomThomson for looking into this! Very appreciated, so as Canvas supports by value visualizations and this is a bug that the presentation team will fix @VladLasitsa let's not hide the Try in Lens context panel from canvas

VladLasitsa commented:
@MichaelMarcialis, I fixed some remarks which related to 'convert to lens' functionality and answered on rest ones. Could you please re-review?

VladLasitsa commented:
@elasticmachine merge upstream

VladLasitsa commented:
@stratoula, Could you please re-review again?

stratoula reviewed (commented):
Great Vlad, some more comments from me:
- @MichaelMarcialis do we want it first or second?
<img width="278" alt="image" src="https://user-images.githubusercontent.com/17003240/206172202-c297caed-e04b-4c74-b6e9-66e8d3900d00.png">

- Visualizations built on saved searches are not converted correctly (the dataview selection is correct but I don't see the filters being populated)

VladLasitsa commented on x-pack/test/functional/apps/lens/open_in_lens/dashboard/dashboard.ts:
```diff
@@ -0,0 +1,59 @@
+    before(async () => {
+      await dashboard.initTests();
+    });
+
```
done

VladLasitsa commented on x-pack/plugins/lens/public/types.ts:
```diff
@@ -248,6 +248,10 @@ export type VisualizeEditorContext<T extends Configuration = Configuration> = {
   originatingApp?: string;
   isVisualizeAction: boolean;
+  searchQuery: Query;
+  searchFilters: Filter[];
```
right, fixed

VladLasitsa commented on x-pack/plugins/lens/public/app_plugin/lens_top_nav.tsx:
```diff
@@ -200,6 +209,40 @@ function getLensTopNavConfig(options: {
     });
   }
+
+  if (showReplaceInDashbord) {
```
done

VladLasitsa commented on x-pack/plugins/lens/public/app_plugin/app.tsx:
```diff
@@ -414,6 +420,18 @@ export function App({
   }, []);
 
+  const returnToOriginSwitchLabelForContext =
+    (initialContext?.originatingApp === 'dashboards' ||
```
hm, now I using special flag from context

VladLasitsa commented:
> * Visualizations built on saved searches are not converted correctly (the dataview selection is correct but I don't see the filters being populated)

Fixed.

ThomThomson commented on src/plugins/dashboard/public/application/lib/build_dashboard_container.ts:
```diff
@@ -109,9 +109,11 @@ export const buildDashboardContainer = async ({
+            ? {
+                ...originalPanelState.explicitInput,
+              }
+            : { hidePanelTitles: originalPanelState.explicitInput.hidePanelTitles }),
```
I'm curious about this change, as the editor is in charge of returning the full input, and `hidePanelTitles` is part of the input, you could restore it by ensuring that the value isn't deleted between the Dashboard -> TSVB -> Lens navigation.

This change likely works okay, but it seems a little poorly defined from the code side to say that the `hidePanelTitles` should never update on `save and return` if the new embeddable's type is different.

If we eventually allowed editing of the title and the `hidePanelTitles` key from inside the editor, we would likely expect changes there to be applied, not overwritten.

If at all possible, I would encourage you to try to maintain the `hidePanelTitles` value instead.

VladLasitsa replied:
In that case I will need to put this setting in navigate to lens config, then get this when user click on "Replace in dashboard" and define it in new embeddable input on Lens side. For me it looks strange that Lens will know about specific dashboard panel setting  - `hidePanelTitles` and only when we use conversion logic

ThomThomson replied:
These are weird settings to be honest, `hidePanelTitles`, and `title` are dashboard implementation details, but they are stored in `explicitInput` which means they should be managed on the individual embeddable level. It isn't ideal, but that is how the architecture is at the moment.

That said, I'm okay if you want to leave this here in order to make the implementation on your side simpler, but it would be nice to have a comment at least explaining that this is temporary, and only required because the key is stored in `explicitInput` when it should be stored outside of it instead.

On our side, we could look into moving this to a better place.

VladLasitsa replied:
Added comment for that

ThomThomson commented on src/plugins/embeddable/public/lib/panel/panel_header/panel_options_menu.tsx:
```diff
@@ -47,15 +55,29 @@ export class PanelOptionsMenu extends React.Component<PanelOptionsMenuProps, Sta
     this.setState({ actionContextMenuPanel: undefined });
     const actionContextMenuPanel = await this.props.getActionContextMenuPanel();
+    const showNotification = actionContextMenuPanel.actions.some(
+      (action) => action.showNotification
```
Nice generic addition to have the action register a `showNotification` boolean! :fire:

ThomThomson reviewed (commented):
@VladLasitsa, I've looked through your proposed solution to the Canvas saved Object ID bug, and I think it will work great with one small change: 

I think it needs to match the Dashboard incoming embeddable implementation, where if the `type` is different, we don't spread **any** of the `originalInput` at all. If you're comfortable with making that fix in the Canvas app, feel free to do so and I'll re-review. Otherwise, we can leave this as a bug for now and fix it on our side! 

Thanks for looking into this!

I've also left a few comments on the Presentation team changes.

MichaelMarcialis commented:
Thanks for those replies, @VladLasitsa! Some additional comments below. 

> > Can we change the "Convert to Lens" action in the options context menu to "Convert visualization to Lens" to better match the wireframes?
> 
> Adding 'visualization' word will wrap text on two lines. I don't think that 'visualization' is too required here, it will be understandable for user that this action will convert 'visualization'

The only reason I ask is because newer users may not be familiar with the term "Lens" and may have no idea that it is an alternative visualization editor. Adding the word "visualization" to the action makes it a bit clearer what exactly is being converted. Any chance that we could increase the width of this menu to accommodate this additional word to prevent the line break?

> > Can we add a save icon to the Save and return and Save modal buttons to match wireframes?
> 
> Unfortunately no, because we use the common Save modal component, I am not sure that adding icon in that PR will be right

Thanks for that explanation. It's not a big deal in this case. I'll write up a separate issue for that when I get a moment.

> > Converting a visualization and replacing it doesn't carry forward the original panel's title settings in the conversion (both the title text and whether it was set to be hidden or not). Would it be possible for us to carry over these panel title settings in the conversion and replacement process to avoid bothering the user? Also, if it is possible to carry these panel title settings over, can we also append (converted) to the panel title text automatically (similar to how we do when saving a converted visualization to the library)?
> 
> I find a way to carry title, but didn't find a good solution to carry hidden title setting(

Bummer. Thanks for checking into that for me.

MichaelMarcialis commented:
> * @MichaelMarcialis do we want it first or second?

@stratoula, @VladLasitsa: If possible, I'd like the convert action to appear second, after the edit action.

stratoula commented:
@VladLasitsa could we possibly add a ui counter to this panel option? 
@elastic/kibana-presentation I don't know if you track these clicks but I think it would be a good candidate to track it and see how many users actually click this

stratoula commented:
@MichaelMarcialis I see that @KOTungseth prefers to not have the word visualization, mentioned here https://github.com/elastic/kibana/issues/147131#issuecomment-1340003579

VladLasitsa commented:
> @VladLasitsa could we possibly add a ui counter to this panel option? @elastic/kibana-presentation I don't know if you track these clicks but I think it would be a good candidate to track it and see how many users actually click this

@stratoula, Do you mean counter for click action or if panel has this option?

stratoula commented:
For click action, but maybe also the other one is also a good metric.

VladLasitsa commented:
> @VladLasitsa, I've looked through your proposed solution to the Canvas saved Object ID bug, and I think it will work great with one small change:
> 
> I think it needs to match the Dashboard incoming embeddable implementation, where if the `type` is different, we don't spread **any** of the `originalInput` at all. If you're comfortable with making that fix in the Canvas app, feel free to do so and I'll re-review. Otherwise, we can leave this as a bug for now and fix it on our side!


Put this fix in that PR.

VladLasitsa commented:
> > * @MichaelMarcialis do we want it first or second?
> 
> @stratoula, @VladLasitsa: If possible, I'd like the convert action to appear second, after the edit action.

@MichaelMarcialis, Done

VladLasitsa commented:
> For click action, but maybe also the other one is also a good metric.

@stratoula, Added for click for now

ThomThomson commented on x-pack/plugins/canvas/public/components/hooks/workpad/use_incoming_embeddable.ts:
```diff
@@ -53,11 +53,20 @@ export const useIncomingEmbeddable = (selectedPage: CanvasPage) => {
+        if (originalType !== type) {
+          updatedInput = incomingInput;
+        } else {
+          updatedInput = { ...originalInput, ...incomingInput };
```
Thank you for adding this fix! Now we can close this issue when this PR merges.

https://github.com/elastic/kibana/issues/147032

ThomThomson reviewed (approved):
Looked through again, and everything LGTM! One nit is that it would be nice to have a comment explaining the `hidePanelTitles` change when dealing with the dashboard incoming embeddable. Really cool feature!

VladLasitsa commented:
@elasticmachine merge upstream

VladLasitsa commented:
@elasticmachine merge upstream

stratoula reviewed (approved):
I am super excited about this feature 😍 
LGTM, I tested it both in Canvas and Dashboard and works great! Thanx Vlad

VladLasitsa commented:
@MichaelMarcialis, Could you please review again?

VladLasitsa commented:
@elastic/kibana-global-experience Could you please review?

clintandrewhall reviewed (approved):
Our code owners changes appear minimal.  LGTM!

MichaelMarcialis reviewed (approved):
Thanks for making those changes, @VladLasitsa! This looks great! I've added two minor comments below, but approving now under the assumption that they can be addressed. 

- When converting a by-reference visualization to Lens, can the breadcrumb reference to the saved object name be contained in quotes? My previous comment to remove the quotes in the breadcrumbs was only meant to apply when converting by-value visualizations without a panel title (where we use the visualization type instead of a name, which doesn't need the quotations).

![image](https://user-images.githubusercontent.com/3884767/207147328-e4f4f915-0a14-42f6-848d-ef3fbdec6625.png)

- After saving a converted visualization to the library in Lens and choosing **NOT** to replace the panel on dashboard, users are now given the option to "Save and return" with a `checkInCircleFilled` icon. Per my earlier comments, can you change the text to "Save and replace" with a `save` icon instead?

![image](https://user-images.githubusercontent.com/3884767/207146610-83c0bf49-53e0-42e1-bf1d-9fd6ad48e864.png)

VladLasitsa commented:
> Thanks for making those changes, @VladLasitsa! This looks great! I've added two minor comments below, but approving now under the assumption that they can be addressed.
> 
> * When converting a by-reference visualization to Lens, can the breadcrumb reference to the saved object name be contained in quotes? My previous comment to remove the quotes in the breadcrumbs was only meant to apply when converting by-value visualizations without a panel title (where we use the visualization type instead of a name, which doesn't need the quotations).
> 
> ![image](https://user-images.githubusercontent.com/3884767/207147328-e4f4f915-0a14-42f6-848d-ef3fbdec6625.png)
> 
> * After saving a converted visualization to the library in Lens and choosing **NOT** to replace the panel on dashboard, users are now given the option to "Save and return" with a `checkInCircleFilled` icon. Per my earlier comments, can you change the text to "Save and replace" with a `save` icon instead?
> 
> ![image](https://user-images.githubusercontent.com/3884767/207146610-83c0bf49-53e0-42e1-bf1d-9fd6ad48e864.png)

@MichaelMarcialis thank you, Fixed these small nits
